src/sentry/integrations/pagerduty/integration.py

```diff
@@ -0,0 +1,185 @@
+                },
+                "columnLabels": {"service": "Service"},
+                "mappedColumnLabel": "Sentry Project",
+            }
+        ]
+
+        return fields
+
+    def update_organization_config(self, data):
+
+        if "project_mapping" in data:
+            project_ids_and_services = data.pop("project_mapping")
+
+            with transaction.atomic():
+                PagerDutyServiceProject.objects.filter(
+                    pagerduty_service__in=PagerDutyService.objects.filter(organization_integration=self.org_integration),
+                ).delete()
+
+                for p_id, s in project_ids_and_services.items():
+                    # create the record in the table
+                    project = Project.objects.get(pk=p_id)
+                    service = PagerDutyService.objects.get(id=s["service"])
+                    PagerDutyServiceProject.objects.create(
+                        project=project,
+                        pagerduty_service=service,
+                    )
+
+    def get_config_data(self):
+        config = self.org_integration.config
+        project_mappings = PagerDutyServiceProject.objects.filter(
+            pagerduty_service__in=PagerDutyService.objects.filter(
+                organization_integration_id=self.org_integration.id,
+            ),
+        )
+        data = {}
+        for pm in project_mappings:
+            data[pm.project_id] = {
+                "service": pm.pagerduty_service.id,
+            }
+        config = {}
+        config["project_mapping"] = data
+        return config
+
+    @property
+    def services(self):
+        services = PagerDutyService.objects.filter(
+            organization_integration=self.org_integration,
+        )
+
+        return services
+
+
+class PagerDutyIntegrationProvider(IntegrationProvider):
+    key = "pagerduty"
+    name = "PagerDuty"
+    metadata = metadata
+    features = frozenset([IntegrationFeatures.ALERT_RULE])
+    integration_cls = PagerDutyIntegration
+
+    setup_dialog_config = {"width": 600, "height": 900}
+
+    def get_pipeline_views(self):
+        return [PagerDutyInstallationRedirect()]
+
+    def post_install(self, integration, organization):
+        services = integration.metadata["services"]
+        try:
+            org_integration = OrganizationIntegration.objects.get(
+                integration=integration,
+                organization=organization,
+            )
+        except OrganizationIntegration.DoesNotExist:
+            return
+
+        with transaction.atomic():
+            for service in services:
+                PagerDutyService.objects.create(
+                    organization_integration=org_integration,
+                    integration_key=service["integration_key"],
+                    service_id=service["id"],
+                    service_name=service["name"],
+                )
+
+    def build_integration(self, state):
+        config = json.loads(state.get("config"))
```

Comment:
Where does this JSON come from? Is it coming from the end user?

Reply:
comes from PagerDuty's end in the request https://github.com/getsentry/sentry/pull/14860/files#diff-8e503c1cc06a347194e64485567d7f4bR183